src/App.tsx

```diff
@@ -15,36 +16,120 @@ export const goodsFromServer = [
   'Garlic',
 ];
 
+enum SortType {
+  name = 'name',
+  length = 'length',
+  default = '',
+}
+
+const getVisibleGoods = (
+  goods: string[],
+  sortField: SortType,
+  isReversed: boolean,
+): string[] => {
+  if (!sortField) {
+    return goods;
+  }
+
+  const sortedGoods = [...goods].sort((good1, good2) => {
+    switch (sortField) {
+      case SortType.name:
+        return good1.localeCompare(good2);
+      case SortType.length:
+        return good1.length - good2.length;
+      default:
+        return 0;
+    }
+  });
+
+  if (isReversed) {
+    sortedGoods.reverse();
+  }
+
+  return sortedGoods;
+};
+
 export const App: React.FC = () => {
+  const [sortField, setSortField] = useState<SortType>(SortType.default);
+  const [isReversed, setIsReversed] = useState(false);
+  const [visibleGoods, setVisibleGoods] = useState<string[]>(goodsFromServer);
+
+  useEffect(() => {
+    const sortedVisibleGoods = getVisibleGoods(
+      visibleGoods,
+      sortField,
+      isReversed,
+    );
+
+    setVisibleGoods(sortedVisibleGoods);
+  }, [visibleGoods, sortField, isReversed]);
+
+  const sortAlphabetically = () => {
+    setSortField(SortType.name);
+  };
+
+  const sortByLength = () => {
+    setSortField(SortType.length);
+  };
+
+  const reset = () => {
+    setVisibleGoods(goodsFromServer);
+    setIsReversed(false);
+    setSortField(SortType.default);
+  };
+
+  const reverseList = () => {
+    setIsReversed(!isReversed);
```

Comment:
```suggestion
    setIsReversed((currentIsReversed) =>  !currentIsReversed);
```
It's better to use functional setState in such cases: [updating-state-based-on-the-previous-state](https://react.dev/reference/react/useState#updating-state-based-on-the-previous-state)